### PR TITLE
Ga 13 filter by case access group

### PR DESCRIPTION
**Pull request by sanjayparekh-hmcts**
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)
- [ ] the `keep-helm` label has been added, if the helm release should be persisted after a successful build

Please remove this line and everything above and fill the following sections:

### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/GA-13

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/ccd/domain/model/casedataaccesscontrol/matcher/CaseAccessGroupsMatcher.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import java.util.List;
+import java.util.Optional;
+
+@Slf4j
+@Component
+public class CaseAccessGroupsMatcher implements RoleAttributeMatcher {
+
+    @Override
+    public MatcherType getType() {
+        return MatcherType.ROLENAME;
```
Comment (abul-syed) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/casedataaccesscontrol/matcher/CaseAccessGroupsMatcher.java:24]:
```suggestion
        return MatcherType.CASEACCESSGROUPID;
```

File: src/main/java/uk/gov/hmcts/ccd/domain/model/casedataaccesscontrol/matcher/CaseAccessGroupsMatcher.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    private static final String VALUE_FIELD = "value";
+    private static final String ID_FIELD = "id";
+    private static final String CASE_ACCESS_GROUP_ID_FIELD = "caseAccessGroupId";
+
+    private static final Logger LOG = LoggerFactory.getLogger(CaseAccessGroupsMatcher.class);
```
Comment (abul-syed) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/casedataaccesscontrol/matcher/CaseAccessGroupsMatcher.java:31]:
```suggestion
```
not required, class already annotated with lombok annotation `@Slf4j`

File: src/main/java/uk/gov/hmcts/ccd/domain/model/casedataaccesscontrol/matcher/CaseAccessGroupsMatcher.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+    @Override
+    public boolean matchAttribute(RoleAssignment roleAssignment, CaseDetails caseDetails) {
+        Optional<String> raCaseAccessGroupId = roleAssignment.getAttributes().getCaseAccessGroupId();
+        List<String> caseAccessGroupIds = getCaseAccessGroupIds(caseDetails).orElse(Collections.emptyList());
+        LOG.info("Match role assignment caseAccessGroupId {} with caseAccessGroupIds {} for role assignment {}",
```
Comment (abul-syed) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/casedataaccesscontrol/matcher/CaseAccessGroupsMatcher.java:37]:
```suggestion
        log.info("Match role assignment caseAccessGroupId {} with caseAccessGroupIds {} for role assignment {}",
```
using `@Slf4j`

Comment (abul-syed) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/casedataaccesscontrol/matcher/CaseAccessGroupsMatcher.java:37]:
```suggestion
        log.debug("Match role assignment caseAccessGroupId {} with caseAccessGroupIds {} for role assignment {}",
```

probs best to switch to `debug` level prior to master merge

File: src/main/java/uk/gov/hmcts/ccd/domain/model/casedataaccesscontrol/matcher/CaseAccessGroupsMatcher.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        boolean matched = caseAccessGroupIds.isEmpty() && (ObjectUtils.isEmpty(raCaseAccessGroupId));
+
+        for (String caseAccessGroupId : caseAccessGroupIds) {
+            if (isValuesMatching(raCaseAccessGroupId, caseAccessGroupId)) {
+                matched = true;
+                break;
+            }
+        }
+
+        log.info("Role assignment caseAccessGroupId {} and case details caseAccessGroupIds {} match {}",
```
Comment (abul-syed) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/casedataaccesscontrol/matcher/CaseAccessGroupsMatcher.java:50]:
```suggestion
        log.debug("Role assignment caseAccessGroupId {} and case details caseAccessGroupIds {} match {}",
```
same comment as above - `debug` level

File: src/main/java/uk/gov/hmcts/ccd/domain/model/casedataaccesscontrol/matcher/CaseAccessGroupsMatcher.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+public class CaseAccessGroupsMatcher implements RoleAttributeMatcher {
+
+    @Override
+    public MatcherType getType() {
+        return MatcherType.ROLENAME;
+    }
+
+    private static final String VALUE_FIELD = "value";
+    private static final String ID_FIELD = "id";
+    private static final String CASE_ACCESS_GROUP_ID_FIELD = "caseAccessGroupId";
```
Comment (abul-syed) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/casedataaccesscontrol/matcher/CaseAccessGroupsMatcher.java:29]: should the static fields be moved to the interface?

**Review by abul-syed — CHANGES_REQUESTED** [review]
few changes

**Review by helenbirdhmcts — APPROVED** [review]
Looks good, approved pending tests added by QA

**Review by RebeccaBaker — APPROVED** [review]

**Pull request merged**